File: frontend/src/types/launchpad.ts
export type LaunchpadStartup = {
  id: string;
  name: string;
  slug?: string | null;
  logo_url?: string | null;
  traction_score?: number | null;
  wallet_pubkey?: string | null;
};

export type LaunchpadInvestor = {
  id: string;
  name: string;
  firm?: string | null;
  wallet_pubkey?: string | null;
};

export type LaunchpadMilestone = {
  id: string;
  funding_round_id: string;
  milestone_index: number;
  threshold_score: number;
  release_percent: number;
  released_at?: string | null;
  tx_hash?: string | null;
};

export type LaunchpadTransaction = {
  id: string;
  funding_round_id?: string | null;
  transaction_hash: string;
  block_hash?: string | null;
  contract_uref?: string | null;
  amount_motes?: number | null;
  action: string;
  status: string;
  created_at?: string | null;
};

export type LaunchpadRound = {
  id: string;
  startup_id: string;
  investor_id: string;
  amount_cspr: number;
  escrow_contract_uref: string;
  safe_nft_mint_hash?: string | null;
  status: string;
  created_at: string;
  updated_at?: string | null;
  startup?: LaunchpadStartup | null;
  investor?: LaunchpadInvestor | null;
  milestones?: LaunchpadMilestone[];
  on_chain_transactions?: LaunchpadTransaction[];
};

export type CreateLaunchpadRoundInput = {
  startup_id: string;
  investor_id: string;
  amount_cspr: number;
  milestones: Array<{
    threshold_score: number;
    release_percent: number;
  }>;
};

File: backend/models/launchpad.py
from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import datetime


class MilestoneCreate(BaseModel):
    threshold_score: int = Field(..., ge=0, le=100, description="Traction score threshold (0–100)")
    release_percent: float = Field(..., ge=0, le=100, description="Percent of escrow to release")


class FundingRoundCreate(BaseModel):
    startup_id: str
    investor_id: str
    amount_cspr: float = Field(..., gt=0)
    milestones: List[MilestoneCreate]


class MilestoneResponse(BaseModel):
    id: str
    funding_round_id: str
    milestone_index: int
    threshold_score: int
    release_percent: float
    released_at: Optional[datetime] = None
    tx_hash: Optional[str] = None


class FundingRoundResponse(BaseModel):
    id: str
    startup_id: str
    investor_id: str
    amount_cspr: float
    escrow_contract_uref: str
    safe_nft_mint_hash: Optional[str] = None
    status: str
    created_at: datetime
    updated_at: Optional[datetime] = None


class EvaluateResponse(BaseModel):
    released: bool
    current_score: int
    round_id: str
    milestones_released: List[str] = []
    message: str


class OnChainTransactionResponse(BaseModel):
    id: str
    funding_round_id: Optional[str] = None
    transaction_hash: str
    block_hash: Optional[str] = None
    contract_uref: str
    amount_motes: Optional[float] = None
    action: str
    status: str
    casper_response: Optional[dict] = None
    created_at: datetime
